### PR TITLE
Add Gemma 4 26B LoRA fine-tuning guide

### PR DESCRIPTION
**Pull request by AdonaiVera**
### PR description

Hi! This PR is part of the #TPUSprint initiative. While working with the ecosystem, I noticed that fine-tuning Gemma 4 Instruct 26B on TPUs and using keras was a missing piece, there wasn’t a clear, end-to-end example showing how to actually run it. I adds a complete example and guide for fine-tuning Gemma 4 Instruct 26B on a TPU v5litepod-8.

The 26B model has a few important constraints that are documented in the guide:

* **Memory**
  Must use `bfloat16` policy (not `mixed_bfloat16`). Mixed keeps float32 master weights and exceeds the ~15.75GB/chip HBM limit on v5litepod-8

* **Sequence length**
  MoE activation tensors must be capped at 128 tokens before compile, otherwise HLO temps exceed memory

* **Weight loading**
  Keras ≤ 3.14 has a bug in the sharded H5 loader that can silently drop weights across shards. The example uses a custom `h5py` loader as a workaround, and ofcourse we should use the latest Keras version that contains the backbone of gemma4

* **Saving**
  Only LoRA adapter weights (few MB) are saved, not the full 26B model. Saving is done via an in-memory buffer before uploading to GCS (since `h5py` doesn’t support `gs://` directly)

Tested end-to-end:

* Completed 10 training steps
* Loss decreased as expected
* LoRA weights saved to GCS
* Inference successfully loaded and ran with fine-tuned weights
Screenshot 2026-04-21 at 7 48 08 AM

### CODE REVIEW

File: examples/gemma4_finetuning/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+keras==3.13.2
+keras-hub==0.27.1
+tokenizers==0.22.2
+sentencepiece==0.2.1
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply examples/gemma4_finetuning/requirements.txt:4]: Thanks for the feedback, but h5py and kagglehub are already present in the Kinetic base container, confirmed by a successful end-to-end run. Happy to add them if the maintainer thinks it's worth being explicit.

File: examples/gemma4_finetuning/gemma4_finetuning.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  counts = {"loaded": 0, "skipped": 0}
+
+  def _load_layer(layer, h5_path):
+    if id(layer) in visited:
+      return
+    visited.add(id(layer))
+
+    all_vars = list(getattr(layer, "_trainable_variables", [])) + list(
+      getattr(layer, "_non_trainable_variables", [])
+    )
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply examples/gemma4_finetuning/gemma4_finetuning.py:128]: Um, this was actually intentional. We can’t use Keras `load_weights()` here because of a prefix mismatch bug, when called through the full CausalLM, it prepends `backbone/` to every path and breaks the weight loading.

So we had to reimplement the traversal using `h5py`. Once you do that, the public `trainable_variables` doesn’t work because it’s recursive and returns variables for the whole subtree. We’re walking layer by layer, so we only want the variables for each layer.

There’s no public API in Keras for that, so we use `_trainable_variables`.

File: docs/guides/gemma4_finetuning.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    keras.mixed_precision.set_global_policy("bfloat16")
+    _make_layout_map(keras)
+
+    print("Loading Gemma 4 Instruct 26B weights (~52 GB, this may take several minutes)...")
+    model = keras_hub.models.Gemma4CausalLM.from_preset(
+        "gemma4_instruct_26b_a4b",
+        load_weights=False,
+    )
+    model_path = kagglehub.model_download("keras/gemma4/keras/gemma4_instruct_26b_a4b")
+    _load_sharded_weights(model.backbone, os.path.join(model_path, "model.weights.json"))
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply docs/guides/gemma4_finetuning.md:153]: Good, I'll apply it !, thx

File: examples/gemma4_finetuning/gemma4_finetuning.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+  """
+  import json
+  import pathlib
+
+  import h5py
+  import jax
+  import numpy as np
+  from keras.src.saving import saving_lib
+  from keras.src.saving.keras_saveable import KerasSaveable
+  from keras.src.utils import naming
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply examples/gemma4_finetuning/gemma4_finetuning.py:72]: Umm, the reason I used it, its for the bug that I mentioned before, I'll add a TODO message to improve this when we solve the other bug

File: docs/guides/gemma4_finetuning.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+## GCloud Setup
+
+Authenticate and configure your project:
+
+```bash
+gcloud auth login
+gcloud auth application-default login
+export GOOGLE_CLOUD_PROJECT="your-project-id"
+export GOOGLE_CLOUD_ZONE="us-central1-a"
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply docs/guides/gemma4_finetuning.md:28]: Umm I would like to discus more this. `GOOGLE_CLOUD_PROJECT` is picked up by `capture_env_vars=["GOOGLE_CLOUD_*"]` and sent to the remote worker for GCS access, so replacing it with `KINETIC_PROJECT` could break things.

From other examples, `KINETIC_ZONE` is set alongside `GOOGLE_CLOUD_ZONE`, so the pattern seems to be adding, not replacing.

Happy to add `KINETIC_PROJECT` next to `GOOGLE_CLOUD_PROJECT` if that’s the convention, just want to confirm first.

File: docs/guides/gemma4_finetuning.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+This pattern is covered in depth in the [Environment Variables](env_vars.md) guide.
+
+`keras-hub` and its tokenizer backends are not installed in the Kinetic base container by default. Add a `requirements.txt` to your project so Kinetic picks them up automatically:
+
+```
+keras==3.13.2
+keras-hub==0.27.1
+tokenizers==0.22.2
+sentencepiece==0.2.1
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply docs/guides/gemma4_finetuning.md:62]: Same as the earlier comment on `requirements.txt`, `h5py` and `kagglehub` are already included in the Kinetic base container (confirmed by a full end-to-end run). Happy to add them explicitly if that’s preferred.

File: examples/gemma4_finetuning/gemma4_finetuning.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply examples/gemma4_finetuning/gemma4_finetuning.py:128]: I created this [PR](https://github.com/keras-team/keras/pull/22739) directly in Keras to proposed a solution.

File: examples/gemma4_finetuning/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+keras==3.13.2
```
Comment (jeffcarp) [inline examples/gemma4_finetuning/requirements.txt:1]: Thank you so much for the PR! I'm wondering instead of adding to the requirements file here, can we start making separate directories for each example (starting with this one)? That way each can have its own requirements.txt

Comment (AdonaiVera) [reply examples/gemma4_finetuning/requirements.txt:1]: Oh actually that is a good idea and more scalable too because right now this file is shared with all the examples. What do you think would be a good structure, something like this?

```
examples/
  gemma4_finetuning/
    requirements.txt
    gemma4_finetuning.py
```

If its good let me know, or if you want a different structure I'm happy to change it up and include it in this PR

Comment (jeffcarp) [reply examples/gemma4_finetuning/requirements.txt:1]: That looks perfect, thanks!

Comment (AdonaiVera) [reply examples/gemma4_finetuning/requirements.txt:1]: Awesome, just pushed the changes, moved everything to its own folder, and rebased with the latest changes @jeffcarp, thank you !

File: examples/gemma4_finetuning/gemma4_finetuning.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffcarp) [reply examples/gemma4_finetuning/gemma4_finetuning.py:128]: Thanks for raising this, I'll take a look at the linked PR as well.

File: docs/guides/gemma4_finetuning.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeffcarp) [reply docs/guides/gemma4_finetuning.md:28]: Yes they're not meant to replace each other.

File: docs/guides/gemma4_finetuning.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+
+**Custom weight loading.** The Kaggle preset stores weights across 6 sharded H5 files described by a `model.weights.json` manifest. Keras's built-in `load_weights()` on the full `CausalLM` prepends a `backbone/` prefix that mismatches every path in the manifest. Loading via `model.backbone.load_weights()` avoids that prefix, but Keras ≤ 3.14 has a bug in `ShardedH5IOStore`: after switching to a different shard file, the internal `current_shard_path` pointer is not updated. When a subsequent `keys()` call restores to the stale path, layers whose weights span multiple shards — every MoE expert bank and the token embedding — fail to load, producing a "received 0 variables" error. The solution is to bypass `ShardedH5IOStore` entirely and read the H5 files directly with h5py, pre-sharding each tensor with `jax.device_put` before assigning it to avoid a memory spike on device 0. The complete loader is implemented as `_load_sharded_weights()` in [`examples/gemma4_finetuning.py`](../../examples/gemma4_finetuning/gemma4_finetuning.py).
```
Comment (jeffcarp) [inline docs/guides/gemma4_finetuning.md:79]: Can you add a TODO here to clean this section up when the bug is fixed?

**Review by jeffcarp — APPROVED** [review]
Thank you!

File: docs/guides/gemma4_finetuning.md (hunk 8 above)
Comment (AdonaiVera) [reply docs/guides/gemma4_finetuning.md:79]: Sure, I added a TODO and saw the PR got merged, that is amazing! I’ll make these improvements once the version with that PR is officially released.

**Comment by jeffcarp** [comment]
Thanks so much!
